pallets/oracle/src/lib.rs

```diff
@@ -0,0 +1,511 @@
+			ensure!(!answered_in_round.is_zero(), "AnsweredInRound from oracle is 0");
+			ensure!(answered_in_round.ge(&round_id), "Stale price");
+
+			log::trace!("New price: {}", price);
+
+			// -- Sign using any account
+			Signer::<T, T::AuthorityId>::any_account()
+				.send_unsigned_transaction(
+					|account| OraclePrice {
+						round_id,
+						price,
+						started_at,
+						updated_at,
+						answered_in_round,
+						block_number,
+						public: account.public.clone(),
+					},
+					|payload, signature| Call::store_price { oracle_price: payload, signature },
+				)
+				.ok_or_else(|| "Failed to sign transaction")?
+				.1
+				.map_err(|_| "Failed to send transaction")?;
+
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		/// Validate unsigned call to this module.
+		///
+		/// By default unsigned transactions are disallowed, but implementing the validator
+		/// here we make sure that some particular calls (the ones produced by offchain worker)
+		/// are being whitelisted and marked as valid.
+		fn validate_unsigned(source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			// Firstly let's check that we call the right function.
+			if let Call::store_price { ref oracle_price, ref signature } = call {
+				// ensure it's a local transaction sent by an offchain worker
+				match source {
+					TransactionSource::InBlock | TransactionSource::Local => {},
+					_ => {
+						log::debug!("Not a local transaction");
+						// Return TransactionValidityError˘ if the call is not allowed.
+						return InvalidTransaction::Call.into()
+					},
+				}
+
+				// check public is valid
+				let valid_keys = <FragKeys<T>>::get();
+				log::debug!("Valid keys: {:?}", valid_keys);
+				let pub_key = oracle_price.public.encode();
+				let pub_key: ed25519::Public = {
+					if let Ok(MultiSigner::Ed25519(pub_key)) =
+						<MultiSigner>::decode(&mut &pub_key[..])
+					{
+						pub_key
+					} else {
+						// Return TransactionValidityError if the call is not allowed.
+						return InvalidTransaction::BadSigner.into() // // 问Gio
+					}
+				};
+				log::debug!("Public key: {:?}", pub_key);
+				if !valid_keys.contains(&pub_key) {
+					// return TransactionValidityError if the call is not allowed.
+					return InvalidTransaction::BadSigner.into()
+				}
+
+				let signature_valid =
+					SignedPayload::<T>::verify::<T::AuthorityId>(oracle_price, signature.clone());
+				if !signature_valid {
+					return InvalidTransaction::BadProof.into()
+				}
+				log::debug!("Sending store_price extrinsic");
+				ValidTransaction::with_tag_prefix("PriceFromOracleUpdate")
+					.and_provides(oracle_price.price)
+					// The transaction is only valid for next 5 blocks. After that it's
+					// going to be revalidated by the pool.
+					// TODO param to consider. It could be useful since the price from Oracle is valid only for few hours
+					//.longevity(5)
```

Comment:
hmm we might wanna do some testing on what to do here

Reply:
I removed the longevity since I misread its purpose. The longevity is related to a transaction that is not added into any block for `TransactionValidation::longevity` blocks after being imported to the pool.